erpnext/edi/doctype/common_code/common_code.py

```diff
@@ -60,52 +59,68 @@ def validate_distinct_references(self):
-
-			code_value = code.find(f"./Value[@ColumnRef='{code_column}']/SimpleValue").text
-			code_hash = CommonCode.simple_hash(code_value, 10)
-
-			title = None
-			if title_column:
-				title = code.find(f"./Value[@ColumnRef='{title_column}']/SimpleValue").text
-
-			description = None
-			if description_column:
-				description = code.find(f"./Value[@ColumnRef='{description_column}']/SimpleValue").text
-
-			codes.append(
-				{
-					"name": f"{list_hash}|{idx}|{code_hash}",
-					"code_list": list_name,
-					"common_code": code_value,
-					"title": title,
-					"description": description,
-					"additional_data": content,
-				}
-			)
+	def from_genericode(self, column_map: dict, xml_element: "etree.Element"):
+		"""Populate the Common Code document from a genericode XML element
+
+		Args:
+		    column_map (dict): A mapping of column names to XML column references. Keys: code, title, description
+		    code (etree.Element): The XML element representing a code in the genericode file
+		"""
+		title_column = column_map.get("title")
+		description_column = column_map.get("description")
+
+		title = None
+		if title_column:
+			title = xml_element.find(f"./Value[@ColumnRef='{title_column}']/SimpleValue").text
+
+		description = None
+		if description_column:
+			description = xml_element.find(f"./Value[@ColumnRef='{description_column}']/SimpleValue").text
+
+		self.title = title
+		self.description = description
+		self.additional_data = etree.tostring(xml_element, encoding="unicode", pretty_print=True)
+
+
+def simple_hash(input_string, length=6):
+	return hashlib.blake2b(input_string.encode(), digest_size=length // 2).hexdigest()
+
+
+def get_code_name(code_list: str, code_value: str):
+	return f"{simple_hash(code_list)}|{simple_hash(code_value, 10)}"
+
+
+def import_genericode(code_list: str, file_name: str, column_map: dict, filters: dict | None = None):
+	"""Import genericode file and create Common Code entries"""
+	file_path = frappe.utils.file_manager.get_file_path(file_name)
+	parser = etree.XMLParser(remove_blank_text=True)
+	tree = etree.parse(file_path, parser=parser)
+	root = tree.getroot()
+
+	# Construct the XPath expression
+	xpath_expr = ".//SimpleCodeList/Row"
+	filter_conditions = [
+		f"Value[@ColumnRef='{column_ref}']/SimpleValue='{value}'" for column_ref, value in filters.items()
+	]
+	if filter_conditions:
+		xpath_expr += "[" + " and ".join(filter_conditions) + "]"
+
+	for xml_element in root.xpath(xpath_expr):
+		code_column = column_map["code"]
+		code_value = xml_element.find(f"./Value[@ColumnRef='{code_column}']/SimpleValue").text
+		code_name = get_code_name(code_list, code_value)
+
+		if frappe.db.exists("Common Code", code_name):
+			common_code = frappe.get_doc("Common Code", code_name)
+		else:
+			common_code: "CommonCode" = frappe.new_doc("Common Code")
+			common_code.common_code = code_value
+			common_code.code_list = code_list
+
+		common_code.from_genericode(column_map, xml_element)
+		common_code.save()
```

Comment:
Prohibitively slow? I could't even blink while this imported my test data. How big are your code lists and how many do you want to import?

Reply:
> Prohibitively slow?

It was pretty annoying during local tests, so much so that I took care to make it faster (not by accident or not knowing better). Now you took the same care to make it slower (for other reasons not entirely evident), so I'll have to assume it's going to be "pretty annoying during local tests" again.

Maybe you want to briefly explain these "other reasons"?

Reply:
I usually avoid `bulk_insert`. Yes, it is fast, but the reason for that is that it skips all validations and hooks, making it error prone and hard to customize. I would rather wait a couple seconds instead of skipping all that good stuff. Sure, when a task would take many minutes or hours, we cannot avoid bulk_insert. But for saving a couple of seconds, I think it's not worth it.

Reply:
I have now added a progress bar, so we don't stare at a blank screen during longer imports.

Reply:
I see. Thanks. That makes sense, in principle.

Do we realistically expect the need for validations and hooks on _mass_ import of _Common Codes_ in practice, though? &mdash; I have a bit of a hard time to come to terms with the ask and presumption to deprioritize speed on abstract principle, alone.

The scenario here simply seems a bit far fetched to me.

The progress bar is a good idea to mitigate the problem. At least people wouldn't try to reload the page because it goes unresponsive.

It's been a while and I can't remember exactly how long it took. Granted, "prohibitively", indeed says as much about my personal tolerance level for slowness. 😊 Let me check it again report back, and when it takes more than 5 seconds, we revert back?

Reply:
Freeze screen for about 30 sec:

![image](https://github.com/user-attachments/assets/78490844-eee2-4002-9138-b5f984ab5c9f)


That's not instant, but probably ok.

Reply:
Loading, otoh, was 8 sec. That's totally fine.

Reply:
> Freeze screen for about 30 sec:
> That's not instant (as before), but probably ok.

iirc, deletion was always slow. didn't change anything here.